File: frontend/modules/trigger/infraestructure/persistence/strapi-trigger-repository.ts
import { TriggerRepository } from '../../domain/trigger-repository'
import { Trigger } from '../../domain/trigger'
import { StrapiClient } from '../../../shared/infraestructure/strapi/strapi-client'
import { TriggerModel, triggerModelToAggregate } from '../strapi/trigger-model'
import { TriggerNotFoundError } from '../../domain/trigger-error'

export class StrapiTriggerRepository implements TriggerRepository {

  async findByDiscordServerId(discordServerId: string): Promise<Trigger[]> {
    const client = new StrapiClient()
    const response = await client.request<TriggerModel>('GET', `/triggers?filters[discord_server_id][$eq][0]=${discordServerId}&populate[0]=output_audio`)
    return response.data.map(triggerModelToAggregate)
  }

  async findId(uuid: string): Promise<Trigger> {
    const client = new StrapiClient()
    const response = await client.request<TriggerModel>('GET', `/triggers?filters[uuid][$eq][0]=${uuid}&populate[0]=output_audio`)
    const trigger = response.data.map(triggerModelToAggregate).shift()

    if (!trigger) {
      throw new TriggerNotFoundError(`trigger with uuid ${uuid} not found`)
    }

    return trigger
  }

  save(trigger: Trigger): void {
  }

}

File: context/src/main/kotlin/com/usadapekora/context/trigger/infraestructure/persistence/strapi/StrapiTriggerRepository.kt
package com.usadapekora.context.trigger.infraestructure.persistence.strapi

import com.usadapekora.context.shared.infraestructure.strapi.client.StrapiFilter
import com.usadapekora.context.shared.infraestructure.strapi.client.StrapiRequest
import com.usadapekora.context.trigger.domain.Trigger
import com.usadapekora.context.trigger.domain.TriggerRepository
import com.usadapekora.context.shared.infraestructure.strapi.resources.Collection
import com.usadapekora.context.trigger.domain.TriggerException
import kotlinx.coroutines.runBlocking


class StrapiTriggerRepository : TriggerRepository {

    private fun mapModelCollectionToAggregate(collection: Collection<TriggerModel>): Array<Trigger>
        = collection.data.map {
            trigger ->
            trigger.toAggregate()
        }.toTypedArray()

    override fun all(): Array<Trigger> {
        val response = runBlocking {
            StrapiRequest("triggers").get(populate = arrayOf("output_audio"))
        }

        if (response != null) {
            return mapModelCollectionToAggregate(response.toModelCollection())
        }

        return arrayOf()
    }

    override fun findByDiscordServer(id: Trigger.TriggerDiscordGuildId): Array<Trigger> {
        val response = runBlocking {
            StrapiRequest("triggers").get(
                filters = arrayOf(StrapiFilter("discord_server_id", id.value)),
                populate = arrayOf("output_audio")
            )
        }

        if (response != null) {
            return mapModelCollectionToAggregate(response.toModelCollection())
        }

        return arrayOf()
    }

    override fun save(trigger: Trigger) { }

    override fun find(id: Trigger.TriggerId): Trigger {
        throw TriggerException.NotFound()
    }

    override fun delete(trigger: Trigger) { }
}
